src/utils/src/SerialDispatchQueue.c

```
@@ -0,0 +1,267 @@
+    }
+
+    LEAVES();
+    return retStatus;
+}
+
+/**
+ * Frees and de-allocates the serial dispatch queue
+ */
+STATUS serialDispatchQueueFree(PSERIAL_DISPATCH_QUEUE_HANDLE pHandle)
+{
+    ENTERS();
+    STATUS retStatus = STATUS_SUCCESS;
+    PSerialDispatchQueue pSerialDispatchQueue = NULL;
+
+    CHK(pHandle != NULL, STATUS_NULL_ARG);
+
+    pSerialDispatchQueue = FROM_SERIAL_DISPATCH_QUEUE_HANDLE(*pHandle);
+
+    serialDispatchQueueFreeInternal(&pSerialDispatchQueue);
+
+    // Set the handle pointer to invalid
+    *pHandle = INVALID_TIMER_QUEUE_HANDLE_VALUE;
+
+CleanUp:
+
+    LEAVES();
+    return retStatus;
+}
+
+STATUS serialDispatchQueueShutdown(SERIAL_DISPATCH_QUEUE_HANDLE handle)
+{
+    ENTERS();
+    STATUS retStatus = STATUS_SUCCESS;
+    PSerialDispatchQueue pSerialDispatchQueue = NULL;
+
+    CHK(IS_VALID_SERIAL_DISPATCH_QUEUE_HANDLE_VALUE(handle), STATUS_INVALID_ARG);
+
+    pSerialDispatchQueue = FROM_SERIAL_DISPATCH_QUEUE_HANDLE(handle);
+
+    ATOMIC_STORE_BOOL(&pSerialDispatchQueue->shutdown, TRUE);
+    CVAR_SIGNAL(pSerialDispatchQueue->executorCvar);
+    THREAD_JOIN(pSerialDispatchQueue->executorTid, NULL);
+
+CleanUp:
+
+    LEAVES();
+    return retStatus;
+}
+
+STATUS serialDispatchQueueDispatchTask(SERIAL_DISPATCH_QUEUE_HANDLE handle,
+                                       DispatchQueueTaskFunc dispatchQueueTaskFn,
+                                       UINT64 customData,
+                                       DISPATCH_QUEUE_TASK_PRIORITY priority)
+{
+    ENTERS();
+    STATUS retStatus = STATUS_SUCCESS;
+    PSerialDispatchQueue pSerialDispatchQueue = NULL;
+    BOOL locked = FALSE;
+    PDispatchQueueTask pDispatchQueueTask = NULL;
+
+    CHK(dispatchQueueTaskFn != NULL, STATUS_NULL_ARG);
+    CHK(IS_VALID_SERIAL_DISPATCH_QUEUE_HANDLE_VALUE(handle), STATUS_INVALID_ARG);
+
+    pSerialDispatchQueue = FROM_SERIAL_DISPATCH_QUEUE_HANDLE(handle);
+
+    CHK(!ATOMIC_LOAD_BOOL(&pSerialDispatchQueue->shutdown), retStatus);
```

I think this should return an error

---

fixed